Ignore delinquent stake on exit

michaelh-laine opened the pull request:
#### Problem
https://github.com/solana-labs/solana/issues/18863 mentions some issues with the exit command, this PR addresses one of these issues only: ability to ignore delinquency when exiting

#### Summary of Changes
Add an --ignore delinquent-stake (-i) flag to the solana-validator exit and wait-for-restart-window sub commands which if passed will cause the wait_for_restart_window function to ignore delinquent stake when in a leader schedule hole.

michaelh-laine commented:
@mvines can you have a look at this, it adds a flag to the solana-validator exit and wait-for-restart-window sub commands, the check " ./scripts/cargo-for-all-lock-files.sh stable check --locked --tests --bins --examples" works locally but seems to fail on the CI so not sure where to go from here, any tips?

mvines commented on validator/src/main.rs:
```diff
@@ -2119,6 +2125,13 @@ pub fn main() {
                     .default_value("10")
                     .help("Minimum time that the validator should not be leader before restarting")
             )
+            .arg(
```
How about adding a `--min-delinquency` argument that defaults to 5% instead?  Great idea overall though!

michaelh-laine replied:
should be max though right? Even the current min_delinquency_percentage should probably be max, because this is the maximum delinquency to accept?

mvines commented on validator/src/main.rs:
```diff
@@ -2126,11 +2131,12 @@ pub fn main() {
+                Arg::with_name("max_delinquent_stake")
+                    .long("max-delinquent-stake")
+                    .takes_value(true)
+                    .validator(is_parsable::<u8>)
+                    .value_name("PERCENT")
+                    .help("The maximum delinquent stake % permitted for an exit [default: 5]")
```
```suggestion
                    .validator(solana_clap_utils:: input_validators::is_valid_percentage)
                    .default_value("5")
                    .value_name("PERCENT")
                    .help("The maximum delinquent stake % permitted for an exit")
```

michaelh-laine replied:
I didn't set the default and used value_t!() as this was used elsewhere (e.g. --tpu-coalesce-ms), I guess no need for the constant if it's not used anywhere else, let me adjust

mvines commented on validator/src/main.rs:
```diff
@@ -105,6 +105,8 @@ const INCLUDE_KEY: &str = "account-index-include-key";
 const DEFAULT_MIN_SNAPSHOT_DOWNLOAD_SPEED: u64 = 10485760;
 // The maximum times of snapshot download abort and retry
 const MAX_SNAPSHOT_DOWNLOAD_ABORT: u32 = 5;
+// The maximum delinquent stake to permit for a safe validator exit/restart
+const DEFAULT_MAX_DELINQUENT_STAKE_PERCENT: u8 = 5;
```
```suggestion
```

mvines commented on validator/src/main.rs:
```diff
@@ -150,8 +151,12 @@ fn wait_for_restart_window(
-    if ignore_delinquent_stake {
-        println!("{}", style("Ignoring delinquent stake").yellow());
+
+    if max_delinquency_percentage != DEFAULT_MAX_DELINQUENT_STAKE_PERCENT {
+        println!(
+            "Maximum permitted delinquency: {}%",
+            max_delinquency_percentage
+        );
```
```suggestion
    println!(
        "Maximum permitted delinquency: {}%",
        max_delinquency_percentage
    );
```

michaelh-laine commented:
@mvines check it out, 

- changed to --max-delinquent-stake (no short flag as m is taken by monitor), 

- I couldn't find a way to validate the input range within the argument definition so I've had to use a clunky conditional (twice) further down to prevent a value >100 (though in reality maybe it doesn't even matter whether we validate for this, it doesn't change anything ultimately), 

- I also added DEFAULT_MAX_DELINQUENT_STAKE_PERCENT (as a u8 so it's not the previous decimal value of 0.05, to make user input easier, converting it at the time of comparison seemed easier (and in line with how the supermajority f64 change was done with the mainnet crash a few weeks ago IIRC)).

mvines commented on validator/src/main.rs:
```diff
@@ -2289,10 +2296,18 @@ pub fn main() {
+                    "Maximum permitted delinquency can't exceed 100"
+                );
+                exit(1);
+            }
```
```suggestion
            let max_delinquent_stake = value_t_or_exit!!(subcommand_matches, "max_delinquent_stake", u8);
```

mvines reviewed (commented):
lgtm, just made some comments about how we can do the clap.rs arg parsing a little cleaner

mvines commented on validator/src/main.rs:
```diff
@@ -2119,6 +2126,15 @@ pub fn main() {
+                    .validator(solana_clap_utils:: input_validators::is_valid_percentage)
+                    .default_value("5")
+                    .value_name("PERCENT")
+                    .help("The maximum delinquent stake % permitted for an exit [default: 5]")
```
You don't need the " [default: 5]" here, clap automatically adds it when the `default_value()` arg is present

michaelh-laine replied:
yeah missed that, adjusted now. Should I worry about the CI failures? looks like an RPC test timeout

mvines commented on validator/src/main.rs:
```diff
@@ -2119,6 +2126,15 @@ pub fn main() {
+                Arg::with_name("max_delinquent_stake")
+                    .long("max-delinquent-stake")
+                    .takes_value(true)
+                    .validator(solana_clap_utils:: input_validators::is_valid_percentage)
```
Sorry, just throw `is_valid_percentage` in the global `use` at the top of the file and then this and the other `.validator()` reduce down to:
```suggestion
                    .validator(is_valid_percentage)
```

mvines reviewed (approved):
Thanks for doing this! I'll usher this into master and the v1.7 branch for the next v1.7 release
